streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -648,55 +648,61 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
       for {
         input       <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
         queueReader  = ZChannel.fromInput(input)
-        outgoing    <- Queue.bounded[ZIO[Env1, Unit, Either[OutDone, OutElem2]]](bufferSize)
+        outgoing    <- Queue.bounded[Fiber[Either[Unit, OutDone], OutElem2]](bufferSize)
         _           <- scope.addFinalizer(outgoing.shutdown)
         errorSignal <- Promise.make[Unit, Nothing]
         permits     <- Semaphore.make(n.toLong)
         failure     <- Ref.make[Cause[OutErr1]](Cause.empty)
-        pull        <- (queueReader >>> self).toPullIn(scope)
+        pull        <- (queueReader >>> self).toPullInAlt(scope)
         _ <- pull
                .foldCauseZIO(
                  cause =>
-                   failure.update(_ && cause) *> outgoing.offer(Exit.failCause(Cause.unit)) *> Exit.failCause(
-                     Cause.unit
-                   ),
-                 {
-                   case Left(outDone) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *> outgoing.offer(ZIO.succeed(Left(outDone)))
-                   case Right(outElem) =>
-                     for {
-                       p     <- Promise.make[Unit, OutElem2]
-                       latch <- Promise.make[Nothing, Unit]
-                       _     <- outgoing.offer(p.await.map(Right(_)))
-                       _ <- permits.withPermit {
-                              latch.succeed(()) *>
-                                ZIO.uninterruptibleMask { restore =>
-                                  restore(errorSignal.await).raceFirstAwait(
-                                    restore(f(outElem))
-                                      .catchAllCause(cause => failure.update(_ && cause) *> Exit.failCause(Cause.unit))
-                                  )
-                                }.foldCauseZIO(
-                                  _ => p.refailCause(Cause.unit) *> errorSignal.refailCause(Cause.unit),
-                                  p.succeed
+                   cause.failureOrCause match {
+                     case Left(Left(outErr)) =>
+                       failure.update(_ && Cause.fail(outErr)) *>
+                         outgoing.offer(Fiber.fail(Left(()))) *>
+                         Exit.failCause(Cause.unit)
+                     case Left(Right(outDone)) =>
+                       permits.withPermits(n.toLong)(ZIO.unit).interruptible *> outgoing.offer(
+                         Fiber.fail(Right(outDone))
```

Comment:
I think you can reuse the existing `Right` on the LHS of the match

Reply:
Done